### PR TITLE
feat(console): select

### PR DESCRIPTION
**Pull request by wangsijie**
## Summary

Add form component: Select.

## Linear Issue Reference

LOG-2191

## Testing

image

image

image

### CODE REVIEW

File: packages/console/src/components/Dropdown/DropdownItem.module.scss [hunk 1]
```diff
@@ -1,12 +1,14 @@
 @use '@/scss/underscore' as _;
 
 .item {
-  padding: _.unit(2) _.unit(4);
+  padding: 6px _.unit(2);
```
Comment (simeng-li) [inline packages/console/src/components/Dropdown/DropdownItem.module.scss:4]: nit: mix absolute number and unit looks odd

File: packages/console/src/components/Dropdown/index.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  const overlayRef = useRef<HTMLDivElement>(null);
+
+  const { position, mutate } = usePosition(anchorRef, overlayRef);
+
+  useEffect(() => {
+    if (isOpen) {
+      // Calc position after modal opened.
+      setTimeout(() => {
+        mutate();
+      });
```
Comment (simeng-li) [inline packages/console/src/components/Dropdown/index.tsx:28]: Could move to `ReactModal`'s  onAfterOpen callback params?

File: packages/console/src/components/Dropdown/DropdownItem.module.scss (hunk 1 above)
Comment (demonzoo) [reply packages/console/src/components/Dropdown/DropdownItem.module.scss:4]: It does. If this is by design, then maybe we can consider using `align-items: center` to center align the inner input

File: packages/console/src/components/Dropdown/index.module.scss [hunk 3]
```diff
@@ -0,0 +1,30 @@
+@use '@/scss/underscore' as _;
+
+.content {
+  background: var(--color-on-primary);
```
Comment (demonzoo) [inline packages/console/src/components/Dropdown/index.module.scss:4]: --color-layer1

File: packages/console/src/components/Dropdown/index.module.scss [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  box-shadow: var(--shadow-light-s2);
+  border-radius: _.unit(2);
+  position: absolute;
+  min-width: 180px;
```
Comment (demonzoo) [inline packages/console/src/components/Dropdown/index.module.scss:8]: `min-width` seems unnecessary to me

File: packages/console/src/components/Dropdown/index.module.scss [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+  &:focus-visible {
+    outline-color: transparent;
```
Comment (demonzoo) [inline packages/console/src/components/Dropdown/index.module.scss:11]: Shouldn't we set the focus-visible outline color to something like `--color-focus-variant`?

File: packages/console/src/components/Dropdown/use-position.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import { RefObject, useCallback, useLayoutEffect, useState } from 'react';
+
+type Position = {
+  left: number;
+  top: number;
+  width: number;
```
Comment (demonzoo) [inline packages/console/src/components/Dropdown/use-position.ts:6]: Do we need `width` in position?

File: packages/console/src/components/Dropdown/use-position.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+) {
+  const [position, setPosition] = useState<Position>();
+
+  const updatePosition = useCallback(() => {
+    if (anchorRef.current && overlayRef.current) {
+      const anchor = anchorRef.current.getBoundingClientRect();
+      const overlay = overlayRef.current.getBoundingClientRect();
+      const isTopSide =
+        anchor.y + anchor.height + overlay.height > window.innerHeight - safePadding;
+      const isLeftSide = anchor.x + overlay.width > window.innerWidth - safePadding;
```
Comment (demonzoo) [inline packages/console/src/components/Dropdown/use-position.ts:24]: nit: `isOnTop` and `isOnLeft` would probably sound better

File: packages/console/src/components/Dropdown/use-position.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import { RefObject, useCallback, useLayoutEffect, useState } from 'react';
+
+type Position = {
+  left: number;
+  top: number;
+  width: number;
+};
+
+const safePadding = 20;
+const overlayPadding = 4;
```
Comment (demonzoo) [inline packages/console/src/components/Dropdown/use-position.ts:10]: nit: better add some comments to explain what are `safePadding` and `overlayPadding`

File: packages/console/src/components/Select/index.module.scss [hunk 9]
```diff
@@ -0,0 +1,48 @@
+@use '@/scss/underscore' as _;
+
+.select {
+  background: var(--color-on-primary);
```
Comment (demonzoo) [inline packages/console/src/components/Select/index.module.scss:4]: layer1

File: packages/console/src/pages/SignInExperience/constants.ts [hunk 10]
```diff
@@ -0,0 +1,3 @@
+import { SignInMethods } from '@logto/schemas';
+
+export const signInMethods: Array<keyof SignInMethods> = ['username', 'sms', 'email', 'social'];
```
Comment (demonzoo) [inline packages/console/src/pages/SignInExperience/constants.ts:3]: This seems unnecessary to me. You can directly use `keyof SignInMethods` to get an array of the keys

File: packages/console/src/components/Dropdown/index.module.scss (hunk 4 above)
Comment (wangsijie) [reply packages/console/src/components/Dropdown/index.module.scss:8]: It looks wired if the width is to small.

File: packages/console/src/components/Dropdown/use-position.ts (hunk 6 above)
Comment (wangsijie) [reply packages/console/src/components/Dropdown/use-position.ts:6]: YES, some times we need the popup layer's width equals to the anchor's width, "select" is an example.

File: packages/console/src/components/Dropdown/DropdownItem.module.scss (hunk 1 above)
Comment (wangsijie) [reply packages/console/src/components/Dropdown/DropdownItem.module.scss:4]: It is by design. And it already has `align-items: center;`, I don't think is good to achieve the `6px` padding by other ways, they may have side effects.

File: packages/console/src/components/Dropdown/index.module.scss (hunk 3 above)
Comment (wangsijie) [reply packages/console/src/components/Dropdown/index.module.scss]: updated

File: packages/console/src/components/Dropdown/use-position.ts (hunk 7 above)
Comment (wangsijie) [reply packages/console/src/components/Dropdown/use-position.ts]: updated

File: packages/console/src/components/Dropdown/use-position.ts (hunk 8 above)
Comment (wangsijie) [reply packages/console/src/components/Dropdown/use-position.ts]: updated

File: packages/console/src/components/Select/index.module.scss (hunk 9 above)
Comment (wangsijie) [reply packages/console/src/components/Select/index.module.scss]: updated

File: packages/console/src/pages/SignInExperience/constants.ts (hunk 10 above)
Comment (wangsijie) [reply packages/console/src/pages/SignInExperience/constants.ts]: This file is unnecessary, removed.

File: packages/console/src/components/Dropdown/index.module.scss (hunk 5 above)
Comment (wangsijie) [reply packages/console/src/components/Dropdown/index.module.scss]: Need to remove focus effect here, changed into `outline: none`.

File: packages/console/src/components/Dropdown/index.tsx (hunk 2 above)
Comment (wangsijie) [reply packages/console/src/components/Dropdown/index.tsx]: updated

File: packages/console/src/components/Dropdown/index.module.scss (hunk 4 above)
Comment (demonzoo) [reply packages/console/src/components/Dropdown/index.module.scss:8]: But 180px seems too big to me. The select component will at least have a downside arrow, right? So technically the minimal width would be 20px (icon) + 2*12 (padding) = 44px. Not that bad to me

Comment (demonzoo) [reply packages/console/src/components/Dropdown/index.module.scss:8]: We might have select buttons like this in our app, so I think we can remove the `min-width` and let the component decide its own width by default.
image

In cases that you feel the select should be wider, use a CSS class to set the width would do the job.

File: packages/console/src/components/Dropdown/DropdownItem.module.scss (hunk 1 above)
Comment (demonzoo) [reply packages/console/src/components/Dropdown/DropdownItem.module.scss:4]: Discussed and we have aligned the usage of `_.unit`. A `.5` step would be fine

File: packages/console/src/components/Dropdown/index.tsx [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+  return (
+    <ReactModal
+      shouldCloseOnOverlayClick
+      isOpen={isOpen}
+      style={{
+        content: position
+          ? {
+              left: `${position.left}px`,
+              top: `${position.top + 4}px`,
```
Comment (demonzoo) [inline packages/console/src/components/Dropdown/index.tsx:31]: What is the extra 4px here?

File: packages/console/src/components/Dropdown/index.tsx [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+              width: isFullWidth ? position.width : undefined,
+            }
+          : { visibility: 'hidden' },
+      }}
+      className={styles.content}
+      overlayClassName={styles.overlay}
+      onRequestClose={onClose}
+      onAfterOpen={() => {
+        mutate();
+      }}
```
Comment (demonzoo) [inline packages/console/src/components/Dropdown/index.tsx:41]: nit: `onAfterOpen={mutate}` would be a little bit better

File: packages/console/src/components/Dropdown/use-position.ts (hunk 6 above)
Comment (demonzoo) [reply packages/console/src/components/Dropdown/use-position.ts:6]: ok, but can we move this `width` logic to that specific `select` component?

File: packages/console/src/components/Dropdown/index.tsx (hunk 12 above)
Comment (wangsijie) [reply packages/console/src/components/Dropdown/index.tsx:41]: updated

File: packages/console/src/components/Dropdown/index.tsx (hunk 11 above)
Comment (wangsijie) [reply packages/console/src/components/Dropdown/index.tsx]: updated as `distance` in `usePosition`

File: packages/console/src/components/Dropdown/use-position.ts (hunk 6 above)
Comment (wangsijie) [reply packages/console/src/components/Dropdown/use-position.ts]: updated

File: packages/console/src/components/Select/index.tsx [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+  const handleSelect = (value: string) => {
+    onChange?.(value);
+    setIsOpen(false);
+  };
+
+  useLayoutEffect(() => {
+    if (anchorRef.current) {
+      const { width } = anchorRef.current.getBoundingClientRect();
+      setWidth(width);
```
Comment (demonzoo) [inline packages/console/src/components/Select/index.tsx:36]: Do we need this useLayoutEffect and setState? Can we directly use the `ref.current` as the width?

**Review by demonzoo — APPROVED** [review]

File: packages/console/src/components/Select/index.tsx (hunk 13 above)
Comment (wangsijie) [reply packages/console/src/components/Select/index.tsx:36]: Interesting, but then it may be executed before render (dom ready)?

**Review by demonzoo — CHANGES_REQUESTED** [review]
It seems you have mixed other contents into this PR?

**Comment by wangsijie** [comment]
@demonzoo fixed with force-push.

**Review by demonzoo — APPROVED** [review]

File: packages/console/src/components/Select/index.tsx (hunk 13 above)
Comment (demonzoo) [reply packages/console/src/components/Select/index.tsx:36]: Let me try it

Comment (wangsijie) [reply packages/console/src/components/Select/index.tsx]: Per offline discussion, changed to the initial implementation.

**Pull request merged**